add multi-track property editor / batch tag editor

ronso0 opened the pull request:
I have been wanting this desperatly for ages for unifying track comments in Mixxx because the alternative workflows are a PITA, e.g. copy/paste per track, or use a tag editor and reload tags in Mixxx.

Intended to fix #9023

This is a modded/stripped clone of DlgTrackInfo, just without DlgTagFetcher and the BPM tab.
It's is not polished yet and I'll push fixes, so please don't take a too close look at the code, yet ; )

* for editable fields <various>
* Cover show, clear and set works (shown empty if covers differ, maybe we can show a special placeholder)
* Star rating works (is 0 if values differ)

### Currently supported properties
**Editable fields:**
- [x] title
- [x] artist
- [x] album title
- [x] album artist
- [x] genre
- [x] grouping (?)
- [x] comment
- [x] track color (rainbow button for multiple track colors)
- [x] musical key
- [x] cover art
- [x] star rating

**Static labels:**
- [x] BPM: show range of loaded track BPM, or nothing if all are 0 / invalid
- [x] bitrate: show common value or range
- [x] duration: show common value or range
- [x] directory: show common directory or _<various>_ (might be helpful to grab info like album title, catalog id etc.)
- [x] "Open in File Browser" button is enabled if all tracks are in the same dir

**Special fields, to be discussed:**
* BPM: I only see few valid uses cases for setting the BPM of multiple tracks:
  * BPM are all horribly off, and scaling via track menu is not applicable for some reason
  * I sometimes explicitly set BPM to `1` (and lock) in order to 'flag' tracks as 'beatless' (athmo, fields recordings and alike) when I'm too lazy to add this to the comments.

**Nice to have:**
* placeholder for [multiple covers] ?
* reset button for each field? (quite some work, probably not worth it, there's the 'Reload from File' button)

ronso0 committed:
add multi-track property editor DlgTrackInfoMulti

ronso0 commented:
Wooh, that was easier than I expected when I took the first look a while back (actually half the time I spent on debugging stupid mistakes..)

This seems to work well for all fields, except cover reload.
(and except key and BPM, I pushed those back since I don't consider them useful here anyway)

fwcd commented:
Wow, very nice to see progress on this. Got to play around with the PR once I find some time.

acolombier commented:
@ronso0 is this ready for review? I can see couple of `TODO`  and commented code. Is there some clean-up to be expected before starting review/testing?

ronso0 commented:
Thanks for taking a look!

I did remove commit with (cover) trace logging and temp. stuff. There are small spots I _may_ polish (turn DEBUG_ASSERT into VERIFY_OR_DEBUG_ASSERT) but that doesn't matter much. For further changes I'll push small commits and maybe squash them later on.

ronso0 commented:
Currently rating and color were stored immediately (not saved to tracks of course).
I'm about to change that so all properties are handled consistently, i.e. apply all changes to track records only when saving.

ronso0 commented:
Please let me know when someone starts reviewing.

ronso0 committed:
Merge remote-tracking branch 'mixxx/main' into trackprop-multi-track

acolombier reviewed (commented):
I just got a chance to test this feature. Overall, everything works fine, although there seems to remain some warning noise when editing fields.

Only got one issue with the current state: differing field separated by `;` are misleading as it get's unclear what happens if you change a "field" (text between two `;` (disclaimer, it updates all tracks field with the summary, now got 10 tracks with the title `My name #1;My name #2;My name #3;My name #4;My name #5;My name #6;My new name #7;My name #8;My name #9;My name #10` :sweat_smile: ). Perhaps, you could make this field read-only? Or make so when it gets edited, the summary disappear (like a placeholder), so it gets clear that all field will be overwritten with the new value. Finally, you could get the individual value. contained between each `;`, although probably overkill and counter-intuitive.

ronso0 commented:
Thanks for testing!

> there seems to remain some warning noise when editing fields.

What are you referring to?

ronso0 commented:
Re: multi-valued fields
I thought, as first implementation, the italic/regular font would be sufficient to signal the IS and NEW (edit) state.
I don't know which separator to choose that is _not_ contained in any tags in order to separate the edited multi-value content and assign it to the correct tracks. 

Ideally, differing values would be indicated and for each field (not directly editable) there would be a dropdown with all available values. (like in Kid3 for example)

We should keep in mind that the main purpose of this batch editor (IMHO) is to assign common value(s) to certain field(s), not edit individual fields per track (that can be done in the library ,or with the single-track editor).

ronso0 commented:
> There was some log warning with just . in the line, not sure where this comes from but looks like that might not have been introduced with your PR so I guess this is out of scope.

Hmm, this might be debug output related to window switching, see #13030

ronso0 commented:
> How about using kVarioustext + italic here too if there is multiple value? During my test, this field was very useful anyway as the value was too large to be displayed. Since you are using this value for RO attributes, perhaps that could be a way to keep things consistent.

Sounds reasonable. I'll try that.
_various_ + combobox with all options would be cool.

ronso0 committed:
DlgTrackInfoMulti: typos and small fixes

ronso0 committed:
DlgTrackInfoMulti: use comboboxes for tag fields

ronso0 commented:
> _<various>_ + combobox with all options would be cool.

This required some fiddling with focus events, but it works :tada: 
I hope UX is good now. I tested all combinations of the belwo steps and didn't encounter issues.

**When testing please also:**
* clear _<various>_ (should clear field for all tracks)
* restore _<various>_ manually (should restore the italic, dim _<various>_ placeholder)
* Reset button
* Reload from file tags

ronso0 commented:
Only unhandled field is now Key.
Will take a look.

ronso0 commented:
I removed the hackish way for clearing multi-value boxes via events etc. There were some situations where restoring values sfrom files didn't work correctly.

This now simply uses  a dedicated  Clear item and everything works as desired.

Only TODO left is adding the key validation when users enter a key text manually.

acolombier commented:
Just had a quick look and it looks much better now, thank you!

Let me know when it is ready for testing.

ronso0 commented:
It's already ready for testing, I only need to implement the key validation.

ronso0 committed:
DlgTrackInfoMulti: add key validation

daschuer commented on src/library/dlgtrackinfomulti.cpp:
```diff
@@ -0,0 +1,833 @@
+}
+
+void DlgTrackInfoMulti::connectTracksChanged() {
+    for (auto it = m_pLoadedTracks.constBegin();
+            it != m_pLoadedTracks.constEnd();
+            it++) {
+        const auto pTrack = it.value();
```
```suggestion
    for (const auto& pTrack : std::as_const(m_pLoadedTracks)) {
```

daschuer reviewed (changes requested):
Some first comments

daschuer commented:
Just tested and it works nice. Thank you very much. 
Does it make sense to gray out the combo box when there is no value to select?

daschuer commented:
This happens after picking a new cover for two tracks: 
```
critical [Main] DEBUG ASSERT: "!m_pLoadedTracks.isEmpty()" in function void DlgTrackInfoMulti::slotCoverFound(const QObject*, const CoverInfo&, const QPixmap&) at ./src/library/dlgtrackinfomulti.cpp:797
```

ronso0 commented:
> Does it make sense to gray out the combo box when there is no value to select?

Not sure what you mean. If there is only one value the combobox is "disabled" in the sense that there is no drop-down.
If a field is empty for all tracks users may want to fill it.

ronso0 commented on src/library/dlgtrackinfomulti.cpp:
```diff
@@ -0,0 +1,833 @@
+}
+
+/// Check if the text has been edited, i.e. is not <various>
+std::optional<QString> validEditText(QComboBox* pBox) {
```
Oh, thanks for the hint!
I vaguely remembered there issues with isNull() / isEmpty() in other code so I didn't consider it here.

daschuer commented:
It was just out of my disappointment to click on the combo box button and nothing happens. No strong demands. 

I have actually a bigger demand to have a combo box with all Genres in the library, to avoid issue like "Hip Hop" vs. "Hiphop" vs "Hip-Hop" or such. 
Or just offer this list:
https://en.m.wikipedia.org/wiki/List_of_ID3v1_genres

ronso0 committed:
DlgTrackInfoMulti: use QString instead of std::optional<QString>, char -> QChar

ronso0 commented:
> It was just out of my disappointment to click on the combo box button and nothing happens.

I understand.
Unfortunately there is no (official) way to disable the drop-down arrow, only the entire combobox (no edit possible). I briefly looked into making the combobox appear as QLineEdit by removing the drop-down button via stylesheet (no icon, zero width) but that doesn't work.

Ideally we'd dynamically switch between QLineEdit (no value or common value) and QComboBox (multiple values).

For now the initially empty box vs. _<various>_  vs. _Text value_ is good enough IMO.

ronso0 commented:
> This happens after picking a new cover for two tracks:
> 
> ```
> critical [Main] DEBUG ASSERT: "!m_pLoadedTracks.isEmpty()" in function void DlgTrackInfoMulti::slotCoverFound(const QObject*, const CoverInfo&, const QPixmap&) at ./src/library/dlgtrackinfomulti.cpp:797
> ```

Can't reproduce. How did you manage to get this? Select a new cover in the dialog, or in the table while the dialog is open?

ronso0 commented on src/library/dlgtrackinfomulti.cpp:
```diff
@@ -0,0 +1,833 @@
+    QList<double> bpms;
+    QList<uint32_t> bitrates;
+    QList<double> durations;
+    QStringList samplerates;
```
Yeah, though that requires some refactoring / more complexity IIUC.
Someting like `convertToQStringConvertible` (actually exactly that) from #13236 would help me a lot I think.

ywwg commented on src/library/dlgtrackinfomulti.cpp:
```diff
@@ -0,0 +1,833 @@
+/// item count to maybe join the list and format the font accordingly.
+/// List can be a QStringList or a QList<double>
+template<typename T>
+void appendToTagList(QList<T>* pList, const T& value) {
```
it is not clear from the function name that this does special management to keep the list unique. Maybe "appendUniqueToTagList" or "addToTagSet" or "includeInTagSet" would make that more clear

ywwg replied:
I agree with daniel's comment that these should be QSets, and then use an adapter to convert them into Lists when needed. it feels brittle otherwise

ronso0 replied:
Jup, that's exactly what I did. Had to deal with some type difficulties but managed it with `convertToQStringConvertible` from #13236

ronso0 replied:
"addToTagSet" sounds good.

ywwg commented on src/library/dlgtrackinfomulti.cpp:
```diff
@@ -0,0 +1,833 @@
+        // No parent because otherwise it inherits the style parent's
+        // style which can make it unreadable. Bug #673411
+        : QDialog(nullptr),
+          m_pUserSettings(std::move(pUserSettings)),
```
std::move does not make sense here because this constructor is called with m_pConfig elsewhere, and that m_pConfig is still owned by the caller, so it cannot be moved into dlgtrackinfomulti.  in general, std::move is used in situations where the object being passed into the function call will *never again* be used by the caller.  This is not the case for the call pattern here, so I believe this std::move will have no effect, and it can be removed.  (UserSettingsPointer is just a shared pointer anyway, it is meant to be copied around and shared, not moved)

ronso0 replied:
I worked off DlgTrackInfo without questioning the basic structure, so I guess it can removed there as well.

daschuer replied:
This does not move the ownership, because it is a shared pointer. It moves the shared pointer to avoid reference counting. So it is a valid approach. 

This follows the pattern to copy objects (in this case the shared pointer) by value parameters and then move them to the final destination.

ywwg replied:
so you're moving the SharedPointer object from the constructor function parameter to the object being created... doesn't the parameter to the constructor go out of scope anyway, so there's no reference counting issue?  it still seems like overkill for such a light object, but ok, we can leave it.

daschuer replied:
I agree that this is micro optimizing this case. 
But we have started to implement this pattern everywhere following the rule of thump.
* const ref for local read access
* value + move when kept as member 
* raw pointer, for write access

It would be interesting if the compiler is able to optimize the reference counting away without the move. I am afraid it cannot. 
The reference counting is not cheap, because memory fences are involved.

ywwg commented on src/library/dlgtrackinfomulti.cpp:
```diff
@@ -0,0 +1,833 @@
+    QList<QComboBox*> valueComboBoxes;
+    valueComboBoxes.append(txtArtist);
+    valueComboBoxes.append(txtTitle);
+    valueComboBoxes.append(txtTitle);
```
title appended twice?

ywwg commented on src/library/dlgtrackinfomulti.cpp:
```diff
@@ -0,0 +1,833 @@
+    // Same procedure for the track color
+    mixxx::RgbColor::optional_t commonColor = m_trackRecords.first().getColor();
+    bool multipleColors = false;
+    for (auto& rec : m_trackRecords) {
```
probably want const auto& since you don't want to modify rec

ywwg commented on src/library/dlgtrackinfomulti.cpp:
```diff
@@ -0,0 +1,833 @@
+    QString commonDir = m_pLoadedTracks.constBegin().value()->getFileInfo().canonicalLocationPath();
+    QStringList dirs;
+    dirs.append(commonDir);
+    for (auto dirIt = m_pLoadedTracks.constBegin();
```
I think std::as_const accomplishes this?

ronso0 replied:
yup, as Daniel mentioned above.

ywwg reviewed (changes requested):
I am excited about this feature!  Do you still plan to add more fixes and address the TODOs for 2.5?

a few notes on C++ style to get started.

ronso0 committed:
DlgTrackInfoMulti: const, use std::as_const QHash iterator

ronso0 committed:
pick util/stringformat.h from #13236

`always_false_v<T>` already exists in midimessage.h, move to new /util/always_false_v.h

ronso0 committed:
DlgTrackInfoMulti: use QSet for temp tag sets

daschuer commented:
> Can't reproduce. How did you manage to get this? Select a new cover in the dialog, or in the table while the dialog is open?

I have selected a new cover from the cover art context menu file picker in the new dialog. It was open for two files.

ywwg commented:
I'll also try actually building/using this.  I also do a lot of track maintenance and this will be super handy

ronso0 commented on src/library/dlgtrackinfomulti.cpp:
```diff
@@ -0,0 +1,835 @@
+template<typename T>
+void DlgTrackInfoMulti::addValuesToComboBox(QComboBox* pBox, QSet<T>& values, bool sort) {
+    // Verify that T can be used for pBox->addItem()
+    DEBUG_ASSERT(!convertToQStringConvertible(*values.constBegin()).isNull());
```
b**shit, will add a dedicated helper.

ronso0 committed:
DlgTrackInfoMulti: add combobox for comment

ywwg commented on src/library/dlgtrackinfomulti.cpp:
```diff
@@ -17,24 +17,24 @@
+        pList->insert(value);
     }
 }
 
```
Now that we are using QSet, you don't need this function at all ("inserts **if value isn't already in the set**, and returns an iterator pointing at the inserted item.") https://doc.qt.io/qt-6/qset.html#insert

ronso0 replied:
thanks, wasn't aware of that. just read that QSet may contain duplicates.

ywwg commented on src/library/dlgtrackinfomulti.cpp:
```diff
@@ -170,22 +172,63 @@ void DlgTrackInfoMulti::init() {
+    connect(txtCommentBox,
+            &QComboBox::currentIndexChanged,
+            this,
+            [this](int idx) {
```
note the clazy warning -- idx not used (so I think you can do `int _` ? )

ronso0 replied:
debug leftover, will remove

ywwg reviewed (commented):
this is cool! I was able to make some bulk changes and it makes sense.  I like "various" for explaining that different tracks already have values set.

One bug, I tried to set the Color of multiple tracks and it did not apply

ronso0 committed:
DlgTrackInfoMulti: fix color apply, remove unused var

ronso0 committed:
DlgTrackInfoMulti: remove obsolete QSet helper function

ronso0 committed:
DlgTrackInfoMulti: don't break/return if re-reading metadata fails

daschuer commented on src/library/dlgtrackinfomulti.cpp:
```diff
@@ -0,0 +1,886 @@
+#include "util/color/color.h"
+#include "util/datetime.h"
+#include "util/duration.h"
+#include "util/optional.h"
```
This can be probably removed?

ronso0 replied:
Yup, removed.

ronso0 commented:
Just a note about an annoying QComboBox bug I noticed:
with long comment lines the popup¹ is much wider than necessary (word-wrap enabled by default).
The QListView and the items have the correct width (max width of the combobox = dialog width) but its _container_ is muuch wider.

It seems the popup width is derived from the box' width hint, i.e. the width of the longest item text, but apparently that doesn't take word-wrap into account. I.e. the multi-line text is displayed correctly but the popup covers the entire screen width. 
I didn't manage to fix it with various elide / sizeAdjust / size policies.

Not sure how to fix this (and I don't expect Qt to fix this soonish). IIUC an custom item delegate wouldn't fix it since the items are actually good. We could sub-class QComboBox and overwrite only the sizeHint / widthHint. I'll do a quick test. If that fails we'll just have to take it as it is.

ronso0 committed:
DlgTrackInfoMulti: don't trim comment text

ronso0 committed:
DlgTrackInfoMulti: minor 'comment' fixes and comments

ronso0 committed:
DlgTrackInfoMulti: limit the width of the comment popup

ronso0 committed:
DlgTrackInfoMulti: move comment combobox above its edit box

ronso0 committed:
DlgTrackInfoMulti: fix some labels' alignment

ronso0 commented:
I managed to work around this bug by setting the max width for the view container, too. It's updated in resizeEvent().
Also I moved the comment combobox up (next to the 'Comments:' label) so it's clear it holds the comments.

This is finally ready to roll I think.

ronso0 commented:
> One bug, I tried to set the Color of multiple tracks and it did not apply

Also fixed, the 'modified' flag was not set.

ronso0 committed:
DlgTrackInfo: fix some labels' alignment

ronso0 committed:
DlgTrackInfoMulti: add File Browser button, enable for common location

daschuer commented:
I was able to reproduce the DEBUG_ASSERT 
Here is the fix: https://github.com/ronso0/mixxx/pull/73

ywwg reviewed (approved):
my notes are addressed, thank you!

ronso0 commented:
> I was able to reproduce the DEBUG_ASSERT Here is the fix: [ronso0#73](https://github.com/ronso0/mixxx/pull/73)

Thanks for your investigation!
I could reproduce with Ok (tried only Apply earlier).

Please correct me if I'm wrong but I think the actual issue is `clear()` which clears the QHash with loaded tracks and the dialog, even though it is now hidden, can still receive the `CoverArtCache::coverFound` signal.

I adopted it from DlgTrackInfo but actually we don't need it at all. These are the relevant cases:
* load tracks from WTrackTableView (via WTrackMenu): the object hold by the unique ptr `m_pDlgTrackInfoMulti` is deleted if tracks are loaded while a DlgTrackInfo/~Multi is open
* closed with X or Cancel: no issue, dialog is deleted by WTrackMenu
https://github.com/mixxxdj/mixxx/pull/12548/files#diff-848ff12a6bbb8c39f28e2a6be33ca2bbbe31f53e6ba550b2f6df7b8bdfaecf1fR2427-R2434
* clicked Okay (with new cover): `CoverArtCache::coverFound` signal is received but who cares, dialog is deleted soon after being closed

ronso0 committed:
DlgTrackInfoMulti: remove clear(), it's not needed

daschuer commented:
I am afraid we need clear(), to release the reference and in turn write the changes to the track file. 
Since there can always be a stray CoverFound signal, we may just deal with it and remove the DEBUG_ASSERT()

ywwg commented:
build warning:

```
AutoUic: /home/owen/src/github/mixxx/src/library/dlgtrackinfomulti.ui: Warning: The name 'horizontalLayout_5' (QHBoxLayout) is already in use, defaulting to 'horizontalLayout_51'.
/home/owen/src/github/mixxx/src/library/dlgtrackinfomulti.ui: Warning: The name 'horizontalSpacer_6' (QSpacerItem) is already in use, defaulting to 'horizontalSpacer_61'.
```

ronso0 commented:
> I am afraid we need clear(), to release the reference ...

Okay, we can do that in the destructor.

> ... and in turn write the changes to the track file.

I don't quite understand?

daschuer commented:
I think the dialog outlives longer, is only hidden after OK. We need to take care that m_pLoadedTrack becomes nullptr, this event will then trigger the write to  the file itself (if this is the last instance) 

Clear() in the destructor us redundant, because it is cleared anyway.

daschuer commented:
Probably a straight forward solution is to just remove the assertion.

ronso0 committed:
DlgTrackInfoMulti: clean up ui file, remove obsolete layouts and spacers

ronso0 committed:
DlgTrackInfoMulti: ensure tracks hash is empty when loading tracks

ronso0 commented:
> I think the dialog outlives longer, is only hidden after OK.

Maybe you overlooked it, I explained when & how the dialog is deleted here https://github.com/mixxxdj/mixxx/pull/12548#issuecomment-2119463777



> We need to take care that m_pLoadedTrack becomes nullptr, this event will then trigger the write to the file itself (if this is the last instance)

I don't understand. Tracks are updated in `saveTracks()` by replacing each TrackRecord. (there is no `TrackPointer m_pLoadedTrack` in `DlgTrackInfoMulti`). 
I though your comment was about explicitly clearing the QHash<TrackId, TrackPointer> (which also happens if the parent dialog is deleted, no?).

I appended a fixup that transforms the DEBUG_ASSERT (which I couldn't trigger aymore with the current state btw).
I also added a demo commit just to `VERIFY_OR_DEBUG_ASSERT(!m_pLoadedTracks.isEmpty())`.

daschuer commented:
Oh, you are right. Now everything works flawlessly. 
Can you rebase the fixup and than we can merge.

ronso0 committed:
DlgTrackInfoMulti: add/remove TODO

ronso0 commented:
Alright.

I noticed some issues I need to fix, though:
* selecting an empty item in the key box does not clear the key
* dialog is still a QTabWidget, though removing the 'Summary' tab miraculously that creates some size policiy / strecth issues with some comboboxes (I love Qt)

ronso0 committed:
DlgTrackInfoMulti ui file: fix names, add blank lines

ronso0 committed:
DlgTrackInfoMulti: remove obsolete QTabWidget

ronso0 committed:
DlgTrackInfoMulti: fix key validation and key reset via empty string

ronso0 committed:
DlgTrackInfoMulti: remove Clear item in a more robust way, commnents

ronso0 committed:
DlgTrackInfoMulti: show more items in the popups

ronso0 committed:
DlgTrackInfoMulti: move comobox change funcs to separate slots

ronso0 commented:
Sooo, the key bug is fixed.

> removing the 'Summary' tab miraculously that creates some size policiy / strecth issues with some comboboxes (I love Qt...)

puuuh, hours of head scratching and fiddling with various SizePolicy combos...
Think I got it right now, at least it somewhat looks like the single-track Track Info dialog.
The "Clear" item is now only an icon, the (X) which is the default Clear icon for the search bar.

I'd appreciate another round of review if someone has the time to do so, and get this in to 2.5-alpha finally.
Thank you!

daschuer commented:
Thank you. Lets merge and test it during beta.
